src/codegen/spirv.zig

```diff
@@ -84,23 +385,197 @@ pub const SPIRVModule = struct {
+            .sub, .subwrap => if (is_float) Opcode.OpFSub else Opcode.OpISub,
+            .mul, .mulwrap => if (is_float) Opcode.OpFMul else Opcode.OpIMul,
+            // TODO: Trap if divisor is 0?
+            // TODO: Figure out of OpSDiv for unsigned/OpUDiv for signed does anything useful.
+            //  => Those are probably for divTrunc and divFloor, though the compiler does not yet generate those.
+            //  => TODO: Figure out how those work on the SPIR-V side.
+            //  => TODO: Test these.
+            .div => if (is_float) Opcode.OpFDiv else if (is_signed) Opcode.OpSDiv else Opcode.OpUDiv,
+            // Only integer versions for these.
+            .bit_and => Opcode.OpBitwiseAnd,
+            .bit_or => Opcode.OpBitwiseOr,
+            .xor => Opcode.OpBitwiseXor,
+            // Int/bool/float -> bool operations.
+            .cmp_eq => if (is_float) Opcode.OpFOrdEqual else if (is_bool) Opcode.OpLogicalEqual else Opcode.OpIEqual,
+            .cmp_neq => if (is_float) Opcode.OpFOrdNotEqual else if (is_bool) Opcode.OpLogicalNotEqual else Opcode.OpINotEqual,
+            // Int/float -> bool operations.
+            // TODO: Verify that these OpFOrd type operations produce the right value.
+            // TODO: Is there a more fundamental difference between OpU and OpS operations here than just the type?
+            .cmp_gt => if (is_float) Opcode.OpFOrdGreaterThan else if (is_signed) Opcode.OpSGreaterThan else Opcode.OpUGreaterThan,
+            .cmp_gte => if (is_float) Opcode.OpFOrdGreaterThanEqual else if (is_signed) Opcode.OpSGreaterThanEqual else Opcode.OpUGreaterThanEqual,
+            .cmp_lt => if (is_float) Opcode.OpFOrdLessThan else if (is_signed) Opcode.OpSLessThan else Opcode.OpULessThan,
+            .cmp_lte => if (is_float) Opcode.OpFOrdLessThanEqual else if (is_signed) Opcode.OpSLessThanEqual else Opcode.OpULessThanEqual,
+            // Bool -> bool operations.
+            .bool_and => Opcode.OpLogicalAnd,
+            .bool_or => Opcode.OpLogicalOr,
+            else => unreachable,
+        };
+
+        try writeInstruction(&self.spv.fn_decls, opcode, &[_]u32{ result_type_id, result_id, lhs_id, rhs_id });
+
+        // TODO: Trap on overflow? Probably going to be annoying.
+        // TODO: Look into SPV_KHR_no_integer_wrap_decoration which provides NoSignedWrap/NoUnsignedWrap.
+
+        if (info.class != .strange_integer)
+            return result_id;
+
+        return self.fail(.{.node_offset = 0}, "TODO: SPIR-V backend: strange integer operation mask", .{});
+    }
+
+    fn genUnOp(self: *DeclGen, inst: *Inst.UnOp) !u32 {
+        const operand_id = try self.resolve(inst.operand);
+
+        const result_id = self.spv.allocResultId();
+        const result_type_id = try self.getOrGenType(inst.base.ty);
+
+        const info = try self.arithmeticTypeInfo(inst.operand.ty);
+
+        const opcode = switch (inst.base.tag) {
```

Comment:
same deal - I think a flatter branching on AIR tags will result in simpler logic, with fewer branches.